src/account.ts

```diff
@@ -0,0 +1,263 @@
+          this.logIn(newAccountUrl);
+          return;
+        }
+
+        // ... otherwise wait until contactUrl of account is loaded
+        accountHandle.on("change", ({ doc }) => {
+          if (doc.contactUrl) {
+            this.logIn(newAccountUrl);
+          }
+        });
+      }
+    });
+  }
+
+  async logIn(accountUrl: AutomergeUrl) {
+    // override old accountUrl
+    localStorage.setItem(ACCOUNT_URL_STORAGE_KEY, accountUrl);
+
+    const accountHandle = this.#repo.find<AccountDoc>(accountUrl);
+    const accountDoc = await accountHandle.doc();
+    const contactHandle = this.#repo.find<ContactDoc>(accountDoc.contactUrl);
+
+    this.#contactHandle.change((oldContact: AnonymousContactDoc) => {
+      if (oldContact.type === "anonymous") {
+        oldContact.claimedBy = contactHandle.url;
+      }
+    });
+
+    this.#contactHandle = contactHandle;
+    this.#handle = accountHandle;
+    this.emit("change");
+  }
+
+  async signUp({ name, avatar }: ContactProps) {
+    let avatarUrl: AutomergeUrl;
+    if (avatar) {
+      avatarUrl = await uploadFile(this.#repo, avatar);
+    }
+
+    this.contactHandle.change((contact: RegisteredContactDoc) => {
+      contact.type = "registered";
+      contact.name = name;
+
+      if (avatarUrl) {
+        contact.avatarUrl = avatarUrl;
+      }
+    });
+  }
+
+  async logOut() {
+    const accountHandle = this.#repo.create<AccountDoc>();
+    const contactHandle = this.#repo.create<ContactDoc>();
+
+    accountHandle.change((account) => {
+      account.contactUrl = contactHandle.url;
+    });
+
+    contactHandle.change((contact) => {
+      contact.type = "anonymous";
+    });
+
+    localStorage.setItem(ACCOUNT_URL_STORAGE_KEY, accountHandle.url);
+
+    this.#handle = accountHandle;
+    this.#contactHandle = contactHandle;
+
+    this.emit("change");
+  }
+
+  get handle() {
+    return this.#handle;
+  }
+
+  get contactHandle() {
+    return this.#contactHandle;
+  }
+}
+
+const ACCOUNT_URL_STORAGE_KEY = "tinyEssayEditor:accountUrl";
+
+let CURRENT_ACCOUNT: Promise<Account>;
+
+async function getAccount(repo: Repo) {
+  if (!repo.storageSubsystem) {
+    throw new Error("cannot create account without storage");
+  }
+
+  if (CURRENT_ACCOUNT) {
+    const currentAccount = await CURRENT_ACCOUNT;
+    if (currentAccount) {
+      return currentAccount;
+    }
+  }
+
+  const accountUrl = localStorage.getItem(
+    ACCOUNT_URL_STORAGE_KEY
+  ) as AutomergeUrl;
+
+  // try to load existing account
+  if (accountUrl) {
+    CURRENT_ACCOUNT = new Promise<Account>(async (resolve) => {
+      const accountHandle = repo.find<AccountDoc>(accountUrl);
+      const contactHandle = repo.find<ContactDoc>(
+        (await accountHandle.doc()).contactUrl
+      );
+      resolve(new Account(repo, accountHandle, contactHandle));
+    });
+
+    return CURRENT_ACCOUNT;
+  }
+
+  // ... otherwise create a new one
+  const accountHandle = repo.create<AccountDoc>();
```

Comment:
eagerly creating anon account docs for every viewer of a doc seems possibly non-ideal for a doc with many viewers, but also seems much simpler than lazy creation when people try to do something, so seems fine